core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java

```diff
@@ -924,7 +925,7 @@ public void start() {
 
                   int concurrentPublishers = Math.min(MAX_INNER_SUBSCRIBERS, targets.size() + (localSegments != null ? 1 : 0));
 
-                  int targetBatchSize = publisher.batchSize / concurrentPublishers + 3;
+                  int targetBatchSize = (publisher.batchSize + concurrentPublishers - 1) / concurrentPublishers;
```

Comment:
> Are you worried that the extra 3 are too many? 

Yes, that's the reasoning. If the user sets batch size to 1 and there's just one publisher then the target batch size should also be 1, not 4.

I'm not sure what you're trying to say about the remainder. The expression I'm using (N + M - 1) / M is equivalent to Math.ceil(N/M), so there's no remainder. E.g. if batch size is 1 and concurrent publishers is 4, (1 + 4 - 1) / 4 = 1.

Reply:
The remainder is say if batchSize = 1 then each publisher will only have 1 (5 + 4 - 1) / 4 = 2, which will end sending 8 entries instead of 5. I agree yours is closer than what we currently, have but was saying if we want it exact we need to take into account the remainder as well and pass that off to only some of the publishers.

Reply:
@wburns you mean if batchSize = 5?
Your proposal is indeed more correct, if you want the sum of targetBatchSizes to be equal to batchSize.
But that's not what I wanted to do here, I just wanted to make the computation of targetBatchSize less surprising, and I think my commit does that.

I was debugging a log that set batchSize=1 and yet CPMI computed targetBatchSize=4, which didn't look correct to me, and it made me read the CPMI code to see what was wrong.
I doubt anyone is going to be surprised when they debug a test with batchSize=5 and they see targetBatchSize=2 in the logs, and it's not like this code is going to make a difference for production code.

Off-topic, I'm not sure we want concurrent publishers at all. We have had too many problems with statistics collectors calling `size()` too often and slowing the cluster down (most recent, https://issues.redhat.com/browse/ISPN-12607), so I would rather prioritize regular reads and writes over publishers.

Reply:
> @wburns you mean if batchSize = 5?
> Your proposal is indeed more correct, if you want the sum of targetBatchSizes to be equal to batchSize.
> But that's not what I wanted to do here, I just wanted to make the computation of targetBatchSize less surprising, and I think my commit does that.

Yeah, I mentioned in my comment that it is indeed better than what we have before, I was just saying if we wanted to make it exact we would need to take another step. I am fine with it this way if this is the furthest we want to go, it is indeed simpler than calculating the remainder.

> Off-topic, I'm not sure we want concurrent publishers at all. We have had too many problems with statistics collectors calling `size()` too often and slowing the cluster down (most recent, https://issues.redhat.com/browse/ISPN-12607), so I would rather prioritize regular reads and writes over publishers.

I am not ready to remove them yet as they have shown considerable increase in performance in both iteration and size calls. The issue more seems just invoking size too many times to begin with. I am not convinced having a single publisher would have helped as it would made size itself probably be 3x slower in which case it would have missed even more caching. Either way this can be discussed separately.

Reply:
> I am fine with it this way if this is the furthest we want to go, it is indeed simpler than calculating the remainder.

Yeah, I wouldn't go further.

> I am not ready to remove them yet as they have shown considerable increase in performance in both iteration and size calls.

I agree that parallelizing makes individual size() calls faster. 
My argument is that if we thought of size() as "slow" instead of "fast, because we do things in parallel", the risks of including it in the statistics would have been obvious from the start, and we wouldn't need caching at all.

Reply:
> My argument is that if we thought of size() as "slow" instead of "fast, because we do things in parallel", the risks of including it in the statistics would have been obvious from the start, and we wouldn't need caching at all.

Sadly, I don't think that thought ever came (even the it is "fast" part), guessing it was more of what should we have and lets put it.